File: JavaScript/StudyPlan/LeetCode75/Q841-KeysAndRooms.js
/**
 * @param {number[][]} rooms
 * @return {boolean}
 */
var canVisitAllRooms = function(rooms) {
    let visitedRoomsCount = 0;
    const totalRooms = rooms.length;
    const visitedRooms = Array(totalRooms);

    function dfs(idx) {
        visitedRooms[idx] = true;
        visitedRoomsCount++;
        return rooms[idx].some(idx => !visitedRooms && dfs(idx));
    }
    dfs(0);
    
    return visitedRoomsCount === totalRooms;
}

File: Python/Leetcode75/Medium/GraphDFS/Q841-KeysAndRooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        n = len(rooms)
        visited = set([0])

        def dfs(room):
            for key in room:
                if key in visited:
                    continue
                visited.add(key)
                dfs(rooms[key])
        
        dfs(rooms[0])

        return len(visited) == n